### ISSUE TITLE
Add something similar to `dash:closedByTypes`

### ISSUE THREAD

**Issue by HolgerKnublauch** [w3c/data-shapes#172]
`sh:closed` is limited to the property shapes at a single node shape, but it does not consider properties that are "inherited" from superclasses.

[`dash:closedByTypes`](https://datashapes.org/constraints.html#ClosedByTypesConstraintComponent) is a solution to that and could be lifted into the SHACL Core vocab. Alternatively, we could add a flag `sh:closedByTypes` to `sh:closed` so that it will also walk up the superclasses.

**Comment by bergos** [comment]
It should also cover `sh:node`. Not sure if it can be the same flag or if someone wants to distinguish between class and node.

Regarding how the flag could be defined, maybe we can allow non-boolean values like `sh:ByType`, `sh:ByNode`, or something like that. Would that count as a breaking change?

**Comment by HolgerKnublauch** [comment]
Yes, good point: `sh:node` is similar to `rdfs:subClassOf` and belongs in the same category.

**Comment by tfrancart** [comment]
See also, in relation with the semantic of `sh:targetClass` #168

**Comment by HolgerKnublauch** [comment]
How do people feel about the design choices: I think if we are adding a new flag sh:closedByType it would be breaking the assumptions of 1.0 tools, because they would flag violations that an 1.2 engine would not.

So I believe this needs to be a new constraint component.

**Comment by bergos** [comment]
We could allow other values instead of adding a new flag.

The new functionality could be used like this:

```turtle
[ sh:closed sh:ByType ]
```

or

```turtle
[ sh:closed sh:ByNode ]
```

The old logic would be still applied with:

```turtle
[ sh:closed true ]
```

or

```turtle
[ sh:closed sh:ByProperties ]
```

**Comment by HolgerKnublauch** [comment]
I have created a PR after several hours of fine tuning the algorithm. It may still not be perfect but the goal is be able to say

sh:closed sh:ByTypes

and have the engine allow any properties that can be reached from the rdf:types of the node including all superclasses, and shapes reachable via sh:targetClass and sh:node.

I didn't copy the example into the core document, and I don't know if leaving it in the test suite is good enough here.

I hope this addresses the main use case with the simplest possible syntax (I automatically made it include rdf:type because that will almost certainly be used by the instances anyway).
